### PR TITLE
add til::color, a universal-converting color type

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
til::color will help us move away from COLORREF internally. It supports
conversion to/from COLORREF, and from all types of structs containing
members named R, G, B and A (or r, g, b, and a).

## Validation Steps Performed
Tests; run through profile/colorScheme deserialization with `til::color`
instead of `uint32_t` or `COLORREF`.

### CODE REVIEW

File: tools/ConsoleTypes.natvis [hunk 1]
```diff
@@ -80,4 +80,8 @@
         <DisplayString Condition="_keyDown">{{↓  wch:{_charData} mod:{_activeModifierKeys} repeat:{_repeatCount} vk:{_virtualKeyCode} vsc:{_virtualScanCode}}</DisplayString>
         <DisplayString Condition="!_keyDown">{{↑ wch:{_charData} mod:{_activeModifierKeys} repeat:{_repeatCount} vk:{_virtualKeyCode} vsc:{_virtualScanCode}}</DisplayString>
     </Type>
+  
+    <Type Name="til::color">
+      <DisplayString>{{RGB: {(int)r,d}, {(int)g,d}, {(int)b,d}; Alpha: {(int)a,d}}}</DisplayString>
```
Comment (zadjii-msft) [inline tools/ConsoleTypes.natvis]: Would VS/Windbg just hate us if we made "Alpha" a literal 'α'?

File: src/inc/LibraryIncludes.h [hunk 2]
```diff
@@ -79,7 +79,9 @@
 #include <wrl.h>
 
 // TIL - Terminal Implementation Library
+#ifndef BLOCK_TIL // Certain projects may want to include TIL manually to gain superpowers
```
Comment (zadjii-msft) [inline src/inc/LibraryIncludes.h]: curiosity: why would someone want this?

File: src/inc/til.h [hunk 3]
```diff
@@ -18,6 +18,111 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
...
+            g{ static_cast<uint8_t>(other.g * 255.0f) },
+            b{ static_cast<uint8_t>(other.b * 255.0f) },
+            a{ static_cast<uint8_t>(other.a * 255.0f) }
+        {
+        }
+
+#ifdef D3DCOLORVALUE_DEFINED
+        constexpr operator D3DCOLORVALUE() const
+        {
+            return D3DCOLORVALUE{ r / 255.0f, g / 255.0f, b / 255.0f, a / 255.0f };
```
Comment (zadjii-msft) [inline src/inc/til.h]: do we need to cast `r`/`g`/`b`/`a` to floats before dividing in C++?

File: src/inc/til.h [hunk 4]
```diff
@@ -18,6 +18,111 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
...
     }
+
+    // color is a universal integral 8bpp RGBA (0-255) color type implicitly convertible to/from
+    // a number of other color types.
+#pragma warning(push)
+    // we can't depend on GSL here (some libraries use BLOCK_GSL), so we use static_cast for explicit narrowing
+#pragma warning(disable : 26472)
+    struct color
+    {
+        uint8_t r, g, b, a;
```
Comment (zadjii-msft) [inline src/inc/til.h]: Should we have a `static_assert` on the size of this type, like `TextColor` does?
```c++
static_assert(sizeof(TextColor) <= 4 * sizeof(BYTE), "We should only need 4B for an entire TextColor. Any more than that is just waste");

```

File: src/inc/til.h [hunk 5]
```diff
@@ -18,6 +18,111 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
...
+            g{ static_cast<uint8_t>((c & 0xFF00) >> 8) },
+            b{ static_cast<uint8_t>((c & 0xFF0000) >> 16) },
+            a{ 255 }
+        {
+        }
+
+        operator COLORREF() const noexcept
+        {
+            return static_cast<COLORREF>(r) | (static_cast<COLORREF>(g) << 8) | (static_cast<COLORREF>(b) << 16);
+        }
```
Comment (zadjii-msft) [inline src/inc/til.h]: What about where we're using `0xffffffff` as `INVALID_COLOR`? How will that work?

File: src/inc/til.h [hunk 6]
```diff
@@ -18,6 +18,111 @@ namespace til // Terminal Implementation Library. Also: "Today I Learned"
...
+#endif
+
+        // Method Description:
+        // - Converting constructor for any other color structure type containing integral R, G, B, A (case sensitive.)
+        // Notes:
+        // - This and all below conversions make use of std::enable_if and a default parameter to disambiguate themselves.
+        //   enable_if will result in an <error-type> if the constraint within it is not met, which will make this
+        //   template ill-formed. Because SFINAE, ill-formed templated members "disappear" instead of causing an error.
+        template<typename TOther>
+        constexpr color(const TOther& other, std::enable_if_t<std::is_integral_v<decltype(std::declval<TOther>().R)> && std::is_integral_v<decltype(std::declval<TOther>().A)>, int> /*sentinel*/ = 0) :
```
Comment (zadjii-msft) [inline src/inc/til.h]: Correct me if I'm wrong, but aren't these al actually only checking for `.R` and `.A`? If I had a type with only those members, what would happen?

File: src/inc/LibraryIncludes.h (hunk 2 above)
Comment (miniksa) [reply src/inc/LibraryIncludes.h]: There's a catch-22 between including TIL in the precomp and also including it late enough in your header include process for it to light up things like D2D color conversions. This is intended to let the minority-case of projects (like DXEngine) block its inclusion and include it later manually so the D2D conversions can light up.

Comment (zadjii-msft) [reply src/inc/LibraryIncludes.h]: okay, that makes sense

File: tools/ConsoleTypes.natvis (hunk 1 above)
Comment (miniksa) [reply tools/ConsoleTypes.natvis]: VS would be fine. WinDBG? Probably fine.

File: src/inc/til.h (hunk 3 above)
Comment (DHowett-MSFT) [reply src/inc/til.h]: it does not appear so

File: src/inc/til.h (hunk 4 above)
Comment (DHowett-MSFT) [reply src/inc/til.h]: I'm not convinced this is necessary... we'd be testing an invariant, and people shouldn't be changing the representation of til::color without a solidly good reason

File: src/inc/til.h (hunk 5 above)
Comment (DHowett-MSFT) [reply src/inc/til.h]: good point: there is no invalid til::color representation. I don't necessarily think that we _should_ have one; that's more of a `std::optional<til::color>`, right?

File: src/inc/til.h (hunk 6 above)
Comment (DHowett-MSFT) [reply src/inc/til.h]: Either the substitution will fail and say there's no constructor, or the constructor will fail and say "there's no member `G`", either of which is fine for us

File: tools/ConsoleTypes.natvis (hunk 1 above)
Comment (DHowett-MSFT) [reply tools/ConsoleTypes.natvis]: but_y.dll

Comment (zadjii-msft) [reply tools/ConsoleTypes.natvis]: all the rest are a single character ¯\\\_(ツ)\_/¯

File: src/inc/til.h (hunk 3 above)
Comment (zadjii-msft) [reply src/inc/til.h]: k

File: tools/ConsoleTypes.natvis (hunk 1 above)
Comment (DHowett-MSFT) [reply tools/ConsoleTypes.natvis]: what I'd really love is a compiled debug extension that shows a color swatch

File: src/inc/til.h (hunk 5 above)
Comment (zadjii-msft) [reply src/inc/til.h]: I suppose, but there are parts of the _console_ codebase particularly that will still need a `0xffffffff` `COLORREF` for `INVALID_COLOR`, that we won't be able to update (I'm thinking esp the propsheet code)

File: src/inc/til.h (hunk 6 above)
Comment (zadjii-msft) [reply src/inc/til.h]: k

File: tools/ConsoleTypes.natvis (hunk 1 above)
Comment (DHowett-MSFT) [reply tools/ConsoleTypes.natvis]: well i made it a literal <insert alpha character here>

**Pull request merged**